interpreter/runtimecost.go

```diff
@@ -0,0 +1,178 @@
+			e.tracker.cost += uint64(math.Ceil(float64(e.tracker.actualSize(args[0], ctx)) * 0.1))
+		case overloads.InList:
+			// If a list is composed entirely of constant values this is O(1), but we don't account for that here.
+			// We just assume all list containment checks are O(n).
+			e.tracker.cost += e.tracker.actualSize(args[1], ctx)
+		// O(min(m, n)) functions
+		case overloads.LessString, overloads.GreaterString, overloads.LessEqualsString, overloads.GreaterEqualsString,
+			overloads.LessBytes, overloads.GreaterBytes, overloads.LessEqualsBytes, overloads.GreaterEqualsBytes,
+			overloads.Equals, overloads.NotEquals:
+			// When we check the equality of 2 scalar values (e.g. 2 integers, 2 floating-point numbers, 2 booleans etc.),
+			// the CostTracker.actualSize() function by definition returns 1 for each operand, resulting in an overall cost
+			// of 1.
+			lhsCost := e.tracker.actualSize(args[0], ctx)
+			rhsCost := e.tracker.actualSize(args[1], ctx)
+			if lhsCost > rhsCost {
+				e.tracker.cost += rhsCost
+			} else {
+				e.tracker.cost += lhsCost
+			}
+		// O(m+n) functions
+		case overloads.AddString, overloads.AddBytes:
+			// In the worst case scenario, we would need to reallocate a new backing store and copy both operands over.
+			e.tracker.cost += e.tracker.actualSize(args[0], ctx) + e.tracker.actualSize(args[1], ctx)
+		// O(nm) functions
+		case overloads.MatchesString:
+			// https://swtch.com/~rsc/regexp/regexp1.html applies to RE2 implementation supported by CEL
+			strCost := uint64(math.Ceil(float64(e.tracker.actualSize(args[0], ctx)) * 0.1))
+			// We don't know how many expressions are in the regex, just the string length (a huge
+			// improvement here would be to somehow get a count the number of expressions in the regex or
+			// how many states are in the regex state machine and use that to measure regex cost).
+			// For now, we're making a guess that each expression in a regex is typically at least 4 chars
+			// in length.
+			regexCost := uint64(math.Ceil(float64(e.tracker.actualSize(args[1], ctx)) * 0.25))
+			e.tracker.cost += strCost * regexCost
+		case overloads.ContainsString:
+			strCost := uint64(math.Ceil(float64(e.tracker.actualSize(args[0], ctx)) * 0.1))
+			substrCost := uint64(math.Ceil(float64(e.tracker.actualSize(args[1], ctx)) * 0.1))
+			e.tracker.cost += strCost * substrCost
+		default:
+			// The following operations are assumed to have O(1) complexity.
+			// 1. Concatenation of 2 lists: see the implementation of the concatList type.
+			// 2. Computing the size of strings, byte sequences, lists and maps: presumably, the length of each of these
+			//    data structures are cached and can be retrieved in constant time.
+			e.tracker.cost += 1
+		}
+	}
+
+	return val
+}
+
+type evalCostTrackerOp struct {
+	Interpretable
+	tracker *CostTracker
+}
+
+func (e *evalCostTrackerOp) Eval(ctx Activation) ref.Val {
+	val := e.Interpretable.Eval(ctx)
+	e.tracker.cost += 1
```

Comment:
What's this for?

Reply:
This is to calculate cost for operation not implemented InterpretableCall [here](https://github.com/google/cel-go/blob/13c5da6dfd824e1a44f5d5e710e1f0882cbade80/interpreter/runtimecost.go#L176). e.g [logic or ](https://github.com/google/cel-go/blob/a29469fda231aa37802d476ef818e2224c25e157/interpreter/interpretable.go#L180).

Reply:
:+1: